packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
+  /// instance, then the method should return true, otherwise it should return
+  /// false. When [oldDelegate] is null, this method should always return true
+  /// unless the new painter initially does not paint anything.
+  ///
+  /// If the method returns false, then the [paint] call might be optimized
+  /// away. However, the [paint] method will get called whenever the
+  /// [RenderEditable]s it attaches to repaint, even if [shouldRepaint] returns
+  /// false.
+  bool shouldRepaint(RenderEditablePainter? oldDelegate);
+
+  /// Paints within the bounds of a [RenderEditable].
+  ///
+  /// The given [Canvas] has the same coordinate space as the [RenderEditable],
+  /// which may be different from the coordinate space the [RenderEditable]'s
+  /// [TextPainter] uses, when the text moves inside the [RenderEditable].
+  ///
+  /// Paint operations performed outside of the region defined by the [canvas]'s
+  /// origin and the [size] parameter may get clipped, when [RenderEditable]'s
+  /// [RenderEditable.clipBehavior] is not [Clip.none].
+  void paint(Canvas canvas, Size size, RenderEditable renderEditable);
+}
+
+class _TextHighlightPainter extends RenderEditablePainter {
+  _TextHighlightPainter({
+      TextRange? highlightedRange,
+      Color? highlightColor
+  }) : _highlightedRange = highlightedRange,
+       _highlightColor = highlightColor;
+
+  final Paint highlightPaint = Paint();
+
+  Color? get highlightColor => _highlightColor;
+  Color? _highlightColor;
+  set highlightColor(Color? newValue) {
+    if (newValue == _highlightColor)
+      return;
+    _highlightColor = newValue;
+    notifyListeners();
+  }
+
+  TextRange? get highlightedRange => _highlightedRange;
+  TextRange? _highlightedRange;
+  set highlightedRange(TextRange? newValue) {
+    if (newValue == _highlightedRange)
+      return;
+    _highlightedRange = newValue;
+    notifyListeners();
+  }
+
+  /// Controls how tall the selection highlight boxes are computed to be.
+  ///
+  /// See [ui.BoxHeightStyle] for details on available styles.
+  ui.BoxHeightStyle get selectionHeightStyle => _selectionHeightStyle;
+  ui.BoxHeightStyle _selectionHeightStyle = ui.BoxHeightStyle.tight;
+  set selectionHeightStyle(ui.BoxHeightStyle value) {
+    assert(value != null);
+    if (_selectionHeightStyle == value)
+      return;
+    _selectionHeightStyle = value;
+    notifyListeners();
+  }
+
+  /// Controls how wide the selection highlight boxes are computed to be.
+  ///
+  /// See [ui.BoxWidthStyle] for details on available styles.
+  ui.BoxWidthStyle get selectionWidthStyle => _selectionWidthStyle;
+  ui.BoxWidthStyle _selectionWidthStyle = ui.BoxWidthStyle.tight;
+  set selectionWidthStyle(ui.BoxWidthStyle value) {
+    assert(value != null);
+    if (_selectionWidthStyle == value)
+      return;
+    _selectionWidthStyle = value;
+    notifyListeners();
+  }
+
+  @override
+  void paint(Canvas canvas, Size size, RenderEditable renderEditable) {
```

Comment:
Could you explain your reasoning in taking TextRange as a parameter and passing RenderEditable into the paint method?  As opposed to maybe dealing with TextBoxes directly.  I don't disagree with the decision, I'm just curious how you designed this at a high level.